src/allotropy/parsers/msd_workbench/msd_workbench_structure.py

```diff
@@ -0,0 +1,168 @@
+from __future__ import annotations
+
+from collections import defaultdict
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Any
+
+import pandas as pd
+
+from allotropy.allotrope.models.shared.components.plate_reader import SampleRoleType
+from allotropy.allotrope.schema_mappers.adm.plate_reader.rec._2024._06.plate_reader import (
+    Measurement,
+    MeasurementGroup,
+    MeasurementType,
+    Metadata,
+)
+from allotropy.exceptions import AllotropeConversionError
+from allotropy.parsers.constants import DEFAULT_EPOCH_TIMESTAMP, NOT_APPLICABLE
+from allotropy.parsers.msd_workbench.constants import (
+    LUMINESCENCE,
+    LUMINESCENCE_DETECTOR,
+    SAMPLE_ROLE_TYPE_MAPPING,
+    SOFTWARE_NAME,
+)
+from allotropy.parsers.utils.pandas import SeriesData
+from allotropy.parsers.utils.uuids import random_uuid_str
+
+
+@dataclass(frozen=True)
+class Header:
+    file_name: str
+    name: str
+    model: str
+
+    @staticmethod
+    def create(file_name: str) -> Header:
+        return Header(
+            file_name=file_name,
+            model=NOT_APPLICABLE,
+            name=SOFTWARE_NAME,
+        )
+
+
+@dataclass(frozen=True)
+class PlateData:
+    measurement_time: str
+    plate_well_count: int
+    well_plate_id: str
+    well_data: list[WellData]
+
+    @staticmethod
+    def create(
+        data: pd.DataFrame,
+    ) -> PlateData:
+        first_row = str(data.iloc[0, 0])
+        if "Plate" not in first_row:
+            msg = "Plate ID not found in the first row of the data"
+            raise AllotropeConversionError(msg)
+        well_plate_id = first_row.split("_")[-1].strip()
+        data = data.iloc[1:].reset_index(drop=True)
+        # Set the first row as the header
+        data.columns = pd.Index(data.iloc[0])
+        data = data[1:].reset_index(drop=True)
+        well_data = []
+        for _row_index, row in data.iterrows():
+            row_series = SeriesData(row)
+            well_data.append(
+                WellData.create(
+                    luminescence=row_series[int, "Signal"],
+                    location_id=row_series[str, "Well"] + "_" + row_series[str, "Spot"],
+                    sample_id=row_series[str, "Sample"] + "_" + row_series[str, "Well"],
+                    concentration=row_series[float, "Concentration"],
+                    measurement_custom_info={
+                        "detection range": row_series.get(str, "Detection Range"),
+                        "assay identifier": row_series.get(str, "Assay"),
+                    },
+                    dilution_factor=row_series.get(int, "Dilution Factor"),
+                )
+            )
+
+        well_column = data["Well"]
+        plate_well_count = len(well_column.unique())
```

Comment:
We should not rely on the raw count here for the plate well count, we should "round up" to a plate well count that we know it correct. See `_get_plate_well_count` in `appbio_quantstudio_data_creator`. You can use similar logic.

It's fine if you even just copy that logic but adapter it to pass in the well ids instead of the `Well` dataclass used there